### ISSUE TITLE
Serialize PKCS#12 CAs with friendly names

### ISSUE THREAD

**Issue by ajoino** [pyca/cryptography#6135]
I'm generating certificates for some systems that use .p12 keystores, and for these systems to work each CA in the keystore needs to have a friendlyName/Alias (I assume they are the same but I am not sure, pkcs12 is confusing...).
I'm using the `pkcs12.serialize_key_and_certificate` function to generate the keystores but I do not know how or if you are able to give the CAs friendly names.

Here's what the CAs I generate look like:
```
Certificate bag
Bag Attributes: <No Attributes>
subject=CN = X

issuer=CN = Y

-----BEGIN CERTIFICATE-----
MII...
```

And here is what they should look like, generated by keytool:
```
Certificate bag
Bag Attributes
    friendlyName: Z
    2.16.840.1.113894.746875.1.1: <Unsupported tag 6>
subject=CN = X

issuer=CN = Y

-----BEGIN CERTIFICATE-----
MII...
```

The simple solution is of course just to use keytool to generate the keystores but I would like to do without that dependency.

**Comment by doug-fitzmaurice-rowden** [comment]
I have the same issue - a 3rd party Java application that parses the P12 file and iterates the contents based on the `aliases()` method. If something doesn't have an alias/friendlyName it doesn't appear.

The `openssl` CLI lets you set the name on a CA using the `-caname` argument. From some investigation this is done using the
`X509_alias_set1()` function:
https://github.com/openssl/openssl/blob/master/apps/pkcs12.c#L634
This puts the alias in an `X509_CERT_AUX` structure, which seems to be an OpenSSL specific set of additions to the ASN1 structure that some functions read/write.

As far as I can see the neither the  `X509_alias_set1` function nor any of the `X509_CERT_AUX` structures are available in `cryptography`, so I don't think is something that I could monkey-patch in either.

**Comment by felixfontein** [comment]
Somewhat related: cryptography now supports (in `main` branch) reading friendly names for all certificates (implemented in #6348).

An earlier version of the PR also allowed to write such PKCS12 files, but I removed that code since the maintainers [didn't want new serialization code for now](https://github.com/pyca/cryptography/pull/6348#issuecomment-933027310). For anyone interested, here's the code to create the test vectors for this, which contains basically the code from the earlier version of the PR: https://gist.github.com/felixfontein/f750763fd5773fbf0ab22af87239aab0 Right now that code won't work out of the box since `X509_alias_set1` isn't exposed by cffi though...

**Comment by reaperhulk** [comment]
If we want to implement this it might be appropriate to write a PKCS12Builder API in the vein of our other builder APIs.

**Comment by ajoino** [comment]
Thanks for the update, I hope you consider adding the builder.

**Comment by ajoino** [comment]
If you accept help from someone who doesn't have experience in contributing to projects like this, I would like to help once I have more time in about two months.

**Comment by ajoino** [comment]
Hi, I started looking into this again, and I have a question.

When serializing keys and certificates to PKCS#12 using `hazmat.primitives.serialization.pkcs12.serialize_key_and_certificates`,
the openssl function `PKCS12_create` in function [`Backend.serialize_key_and_certificates_to_pkcs12`](https://github.com/pyca/cryptography/blob/6327100ae2fc5fef0d30d7eafb1b071972ac4f70/src/cryptography/hazmat/backends/openssl/backend.py#L2214).
According to the [openssl documentation for `PKCS12_create`](https://www.openssl.org/docs/man1.1.1/man3/PKCS12_create.html), the second argument is the friendlyName.
To me, this should mean that the friendlyName gets serialized, but since it doesn't I'm misunderstanding something.
What is it that I am not understanding about `PKCS12_create`?

**Comment by felixfontein** [comment]
@ajoino what do you mean with "but since it doesn't"?

**Comment by felixfontein** [comment]
How do you check that the PKCS#12 file lacks the friendlyName? Also did you provide both a certificate (`cert` parameter) when calling `serialize_key_and_certificates`? Did you also provide a private key for the certificate?

**Comment by ajoino** [comment]
If I seemed confused, it's because I haven't looked at my own code since I created the issue and I've forgot some details, like that this was about the CA friendlyNames and not the cert's... I am no longer confused.

Thanks for the help!

**Comment by ajoino** [comment]
I have been looking into the pkcs12 code trying to figure out what would be the best way to add friendlyNames to the additional certs of serialized pkcs12 certs. As a start, I submitted #6897 to expose `X509_alias_set1`. Before going further, I would like to hear what more experienced contributors and developers have to think. I ignore the possibility of adding another serialization routine for now, as it was explained above that more serialization code is undesirable.

My ideas:
 1. Modify [`serialization.pkcs12.serialize_key_and_certificates`](https://github.com/pyca/cryptography/blob/443cee84f52f5063f014c948068748159056b9c3/src/cryptography/hazmat/primitives/serialization/pkcs12.py#L164) in one of ways
   a. Add an argument `aliases: List[Optional[bytes]]` whose length has to match the length of the `cas` argument. Each element of `aliases` is the friendlyName of the corresponding element in `cas`. `None` means that certificate has no friendlyName. The  function would be modified to add each friendlyName before serialization with a for-loop and zip.
   b. Change the signature of `cas` to `cas: Optional[Iterable[Union[x509.Certificate, PKCS12Certificate]]]`, and add the friendlyName of each PKCS12Certificate to the certificate it contains before serializing. Would also require a single for-loop though with some isinstance checking.
 2. Change `PKCS12Certificate.__init__` to automatically add the `friendly_name` to `cert`.
 3. Based on the comments above, Add `PKCS12Certificate` and `PKCS12KeyAndCertificate` builders. In my head, these would internally contain `CertificateBuilders` and have methods that thinly wraps the contained builder's method, with an additional `add_additional_certs(cert: x509.Certificate, friendly_name: Optional[bytes])` to add friendlyNames to additional certificates.

Additional ideas and feedback are welcome! (Tagging @felixfontein @reaperhulk @doug-fitzmaurice-rowden as you've been involved in the discussion previously)

My thoughts on each option:

Adding builders for pkcs12 would be useful for people who primarily work with that format. However, I think it is heavy-handed for my initial issue, which just concerns serialization.

Modifying the serialization function according to 1a seems like a good solution when you mainly use `x509.Certificate` and rarely serialize to pkcs12, since won't first need to create a `PKCS12Certificate` object. It is backwards incompatible if you don't use named arguments, and would maybe make testing more complex? Need someone with better experience writing tests to comment on that.
1b is fully backwards compatible, afaict, with a cleaner API. However, it does require you to create basically redundant `PKCS12Certificate` objects if you mainly work with `x509.Certificate` objects. It would also make testing more complex, assuming 1a does that.

Lastly, 2 seems like a good idea to do no matter what. If I create a `PKCS12Certificate` with a `friendly_name` it's likely that I would like the certificate to contain that information. However, that might create issues when loading pkcs12 certs (in particular, I don't know what would happen at [this line](https://github.com/pyca/cryptography/blob/443cee84f52f5063f014c948068748159056b9c3/src/cryptography/hazmat/backends/openssl/backend.py#L2126).


To me, 1b seems like the cleanest way to allow serialization of friendlyNames in cas. It requires basically no change to the existing API and behaviour in the currently supported case.

**Comment by ajoino** [comment]
I got impatient over the weekend and implemented solution 1b, see PR #6910.

**Comment by doug-fitzmaurice-rowden** [comment]
Thanks @ajoino, this is some really good work!

I'm not sure I have a useful opinion on the API design, so I'll leave that up to the project maintainers.
I think that 1b (union x509 certs / pkcs cert) is a good choice - and I agree that 3 (builder objects) is a bit complicated for the initial issue.

**Comment by ajoino** [comment]
Most of the credit should go to Felix, I just re-packaged his implementation.

**Comment by felixfontein** [comment]
@ajoino thanks for turning my experiments from back then into a proper PR :)

**Comment by ajoino** [comment]
@felixfontein no I should thank you! Without your pioneering work, I would never have figured out how to do this.

**Comment by alex** [comment]
fixed by #6910
